js/modules/types/message.js

```diff
@@ -166,13 +166,72 @@ const toVersion3 = exports._withSchemaVersion(
 );
 
 // UpgradeStep
-exports.upgradeSchema = async (message, { writeAttachmentData } = {}) => {
-  if (!isFunction(writeAttachmentData)) {
-    throw new TypeError('`context.writeAttachmentData` is required');
+exports.upgradeSchema = async (message, { writeNewAttachmentData } = {}) => {
+  if (!isFunction(writeNewAttachmentData)) {
+    throw new TypeError('`context.writeNewAttachmentData` is required');
   }
 
   return toVersion3(
     await toVersion2(await toVersion1(await toVersion0(message))),
-    { writeAttachmentData }
+    { writeNewAttachmentData }
   );
 };
+
+exports.createAttachmentLoader = (loadAttachmentData) => {
+  if (!isFunction(loadAttachmentData)) {
+    throw new TypeError('`loadAttachmentData` is required');
+  }
+
+  return async message => (Object.assign({}, message, {
+    attachments: await Promise.all(message.attachments.map(loadAttachmentData)),
+  }));
+};
+
+//      createImporter :: (RelativePath -> IO Unit)
+//                        Message ->
+//                        IO (Promise Message)
+exports.createImporter = (writeExistingAttachmentData) => {
+  if (!isFunction(writeExistingAttachmentData)) {
+    throw new TypeError('"writeExistingAttachmentData" must be a function');
+  }
+
+  return async (rawMessage) => {
+    if (!exports.isValid(rawMessage)) {
+      throw new TypeError('"rawMessage" is not valid');
+    }
+
+    const message = exports.initializeSchemaVersion(rawMessage);
+
+    const { attachments } = message;
+    const hasAttachments = attachments && attachments.length > 0;
+    if (!hasAttachments) {
+      return message;
+    }
+
+    const lastVersionWithAttachmentDataInMemory = 2;
+    const willHaveAttachmentsSavedOnFileSystemDuringUpgrade =
+      message.schemaVersion <= lastVersionWithAttachmentDataInMemory;
+    if (willHaveAttachmentsSavedOnFileSystemDuringUpgrade) {
```

Comment:
Very explanatory naming. Much clarity. Many understandings.

Reply:
An slightly shorter name, just food for thought: `willAttachmentsGoToDiskOnUpgrade`

Reply:
LOL! I like `willAttachmentsGoToDiskOnUpgrade` but I struggled with `*Disk*` myself wondering if SSDs wouldn’t count 😜